src/helics/application_api/ValueConverter.cpp

```diff
@@ -29,22 +29,320 @@ template class ValueConverter<std::vector<std::complex<double>>>;
+        data[2] = std::byte((val.size() >> 8U) & 0xFFU);
+        data[3] = std::byte((val.size() & 0xFFU));
+        std::memcpy(data + 4U, val.data(), val.size());
+        return val.size() + 4U;
+    }
+
+    size_t convertToBinary(std::byte* data, const NamedPoint& val)
+    {
+        std::memcpy(data, &val.value, 8);
+        data[8] = npCode | littleEndianCode;
+
+        data[9] = std::byte((val.name.size() >> 16U) & 0XFFU);
+        data[10] = std::byte((val.name.size() >> 8U) & 0XFFU);
+        data[11] = std::byte((val.name.size() & 0XFFU));
+        std::memcpy(data + 12, val.name.data(), val.name.size());
+        return val.name.size() + 12U;
+    }
+
+    size_t convertToBinary(std::byte* data, const std::vector<double>& val)
+    {
+        data[0] = vectorCode | littleEndianCode;
+
+        data[1] = std::byte((val.size() >> 16U) & 0XFFU);
+        data[2] = std::byte((val.size() >> 8U) & 0XFFU);
+        data[3] = std::byte((val.size() & 0XFFU));
+        std::memset(data + 4, 0, 4);
+        std::memcpy(data + 8, val.data(), val.size() * sizeof(double));
+        return val.size() * sizeof(double) + 8U;
+    }
+
+    size_t convertToBinary(std::byte* data, const double* val, size_t size)
+    {
+        data[0] = vectorCode | littleEndianCode;
+
+        data[1] = std::byte((size >> 16U) & 0XFFU);
+        data[2] = std::byte((size >> 8U) & 0XFFU);
+        data[3] = std::byte((size & 0XFFU));
+        std::memset(data + 4, 0, 4);
+        std::memcpy(data + 8, val, size * sizeof(double));
+        return size * sizeof(double) + 8U;
+    }
+
+    size_t convertToBinary(std::byte* data, const std::vector<std::complex<double>>& val)
+    {
+        data[0] = cvCode | littleEndianCode;
+
+        data[1] = std::byte((val.size() >> 16U) & 0XFFU);
+        data[2] = std::byte((val.size() >> 8U) & 0XFFU);
+        data[3] = std::byte((val.size() & 0XFFU));
+        std::memset(data + 4, 0U, 4U);
+        std::memcpy(data + 8, val.data(), val.size() * sizeof(double) * 2);
+        return val.size() * sizeof(double) * 2U + 8U;
+    }
+
+    size_t getDataSize(const std::byte* data, data_type type)
+    {
+        switch (type) {
+            default:
+                return 1;
+            case data_type::helics_complex:
+                return 2;
+            case data_type::helics_string:
+            case data_type::helics_vector:
+            case data_type::helics_complex_vector:
+            case data_type::helics_custom:
+                return (static_cast<size_t>(data[1]) << 16U) +
+                    (static_cast<size_t>(data[2]) << 8U) + static_cast<size_t>(data[3]);
+            case data_type::helics_named_point:
+                return (static_cast<size_t>(data[9]) << 16U) +
+                    (static_cast<size_t>(data[10]) << 8U) + static_cast<size_t>(data[11]);
+        }
+    }
+
+    data_type detectType(const std::byte* data, size_t size)
+    {
+        switch (size) {
```

Comment:
Is there a reason for putting the code mask at the end? Always having the first byte be the code byte would make it easier to detect the type and could enable packing different value types into a single data blob.

Reply:
I think the biggest consideration is keeping the serialization of double and int64 to <=15 bytes, so it fits in the short string optimization size of all platforms and there are no additional allocations for those types.  Additionally complex <=22 bytes.  As that I know is a performance hit if it doesn't fit in a SSO for the std::string.  
(maybe we want to consider a custom container with different Small size optimization?)  

The string pointer is always aligned (that is not the only use case but a very common one)  so having aligned access to copy makes it a single instruction to serialize that part of it.  

I am not sure what the cost of unaligned access is.

Reply:
Actually, the opposite order (footer instead of header) could get rid of the need to add padding by having the data elements first, followed by 4/8 bytes size if needed, and the type of data/endianness at the end. If combining multiple fields into a single packet, null bytes (null type code) could optionally be added as padding to align copies for the subsequent fields, with the downside being that deserializing fields in a class/struct would need to happen in reverse.

I haven't found much on the performance of memcpy for aligned/unaligned on x86. It's mostly embedded systems stuff that appears (https://www.embedded.com/optimizing-memcpy-improves-speed/) on ARM, which probably behaves differently from x86.